File: picking.py
from copy import deepcopy
from sys import argv

from ranking import Ranking

class Picked:
  def __init__(self, size, picked):
    self._picked = [False] * size
    for p in picked:
      self._picked[p] = True

  def next(self, pos):
    pos += 1
    while pos < len(self._picked) and self._picked[pos]:
      pos += 1
    if pos >= len(self._picked):
      return None
    return pos


class Selection:
  def __init__(self, size, max, picked):
    self.selection = [-1] * size
    self._max = max
    self._picked = Picked(max, picked)

  def next(self):
    if self.selection[0] == -1:
      self.selection[0] = self._picked.next(-1)
      self._fill(1)
      return True
    return self._next_pos(len(self.selection)-1)

  def _fill(self, pos):
    for i in range(pos, len(self.selection)):
      self.selection[i] = self._picked.next(self.selection[i - 1])
      if self.selection[i] is None:
        return False
    return True

  def _next_pos(self, pos):
    if pos == -1:
      return False
    self.selection[pos] = self._picked.next(self.selection[pos])
    if self.selection[pos] is None or not self._fill(pos + 1):
      return self._next_pos(pos - 1)
    return True


class Picker:
  def __init__(self, ranking, size, players):
    self._ranking = ranking
    self._size = size
    self._players = players
    self._best_diff = 999999
    self._best = None

  def pick(self, teams, picked):
    if teams == 0:
      best = -999999
      worst = 999999
      for team in picked:
        score = 0
        for player in team:
          score += self._ranking.score(self._players[player])
        if score > best:
          best = score
        if score < worst:
          worst = score
      diff = best - worst
      if diff < self._best_diff:
        self._best = deepcopy(picked)
        self._best_diff = diff
      return
    all_picked = [player for team in picked for player in team]
    selection = Selection(self._size, len(self._players), all_picked)
    while selection.next():
      picked.append(selection.selection)
      self.pick(teams - 1, picked)
      picked.pop()

  def print(self):
    for team in self._best:
      for player in team:
        print(self._players[player], end=' ')
      print()


def main():
  ranking = Ranking()
  ranking.open(argv[1])
  teams = int(argv[2])
  players = []
  for i in range(3, len(argv)):
    players.append(argv[i])
  if len(players) % teams > 0:
    print('Jogadores sobrando: ', len(players) % teams)
    return
  picker = Picker(ranking, len(players) // teams, players)
  picker.pick(teams, [])
  picker.print()

if __name__ == '__main__':
  main()

File: picking.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"sort"
	"strings"
	"strconv"
)

func main() {
	scores := readScores()
	teams, err := strconv.Atoi(os.Args[1])
	if err != nil {
		log.Fatal(err)
	}
	players := os.Args[2:]
	if len(players) % teams > 0 {
		fmt.Printf("Jogadores sobrando: %v\n", len(players) % teams)
		return
	}
	picker := newPicker(scores, len(players) / teams, players)
	picker.pick(teams, nil)
	picker.print()
}

func readScores() map[string]float64 {
	scores := make(map[string]float64)
	r := bufio.NewReader(os.Stdin)
	for true {
		line, err := r.ReadString('\n')
		if err != nil {
			break
		}
		fields := strings.Split(strings.Trim(line, "\n"), " ")
		if fields == nil {
			log.Fatal("fields == nil")
		}
		f, err := strconv.ParseFloat(fields[2], 64)
		if err != nil {
			log.Fatal(err)
		}
		scores[fields[1]] = f
	}
	return scores
}

func newPicker(score map[string]float64, size int, players []string) picker {
	sort.Sort(byScore{players, score})
	return picker{score, size, players, 999999, nil}
}

type picker struct {
	score map[string]float64
	size int
	players []string
	bestDiff float64
	best [][]int
}

func (p *picker) pick(teams int, picked [][]int) {
	if teams == 0 {
		best := -999999.0
		worst := 999999.0
		for _, team := range picked {
			score := 0.0
			for _, player := range team {
				score += p.score[p.players[player]]
			}
			if score > best {
				best = score
			}
			if score < worst {
				worst = score
			}
		}
		diff := best - worst
		if diff < p.bestDiff {
			p.best = make([][]int, len(picked))
			for i := range picked {
				p.best[i] = make([]int, len(picked[i]))
				copy(p.best[i], picked[i])
			}
			p.bestDiff = diff
		}
		return
	}
	var all_picked []int
	for _, team := range picked {
		all_picked = append(all_picked, team...)
	}
	selection := newSelection(p.size, len(p.players), all_picked)
	for selection.next() {
		picked = append(picked, selection.selection)
		p.pick(teams - 1, picked)
		picked = picked[:len(picked) - 1]
	}
}

func (p *picker) print() {
	for _, team := range p.best {
		total := 0.0
		for _, player := range team {
			fmt.Printf("%v ", p.players[player])
			total += p.score[p.players[player]]
		}
		fmt.Printf("%v\n", total)
	}
}

type byScore struct {
	players []string
	score map[string]float64
}

func (b byScore) Len() int {
	return len(b.players)
}

func (b byScore) Swap(i, j int) {
	b.players[i], b.players[j] = b.players[j], b.players[i]
}

func (b byScore) Less(i, j int) bool {
	return b.score[b.players[i]] > b.score[b.players[j]]
}

func newSelection(size, max int, picked []int) *selection {
	s := make([]int, size)
	for i := range s {
		s[i] = -1
	}
	return &selection{s, max, newPicked(max, picked)}
}

type selection struct {
	selection []int
	max int
	picked *picked
}

func (s *selection) next() bool {
	if s.selection[0] == -1 {
		s.selection[0] = s.picked.next(-1)
		s.fill(1)
		return true
	}
	return s.nextPos(len(s.selection) - 1)
}

func (s *selection) fill(pos int) bool {
	for i := pos; i < len(s.selection); i++ {
		s.selection[i] = s.picked.next(s.selection[i - 1])
		if s.selection[i] == -1 {
			return false
		}
	}
	return true
}

func (s *selection) nextPos(pos int) bool {
	if pos == -1 {
		return false
	}
	s.selection[pos] = s.picked.next(s.selection[pos])
	if s.selection[pos] == -1 || !s.fill(pos + 1) {
		return s.nextPos(pos - 1)
	}
	return true
}

func newPicked(size int, pickedList []int) *picked {
	pickedMap := make([]bool, size)
	for _, p := range pickedList {
		pickedMap[p] = true
	}
	return &picked{pickedMap}
}

type picked struct {
	picked []bool
}

func (p *picked) next(pos int) int {
	pos += 1
	for pos < len(p.picked) && p.picked[pos] {
		pos += 1
	}
	if pos >= len(p.picked) {
		return -1
	}
	return pos
}
